src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1094,21 +1100,16 @@
 
             [[(op :guard #{">" "<" ">=" "<="}) column value]]
             (let [{:keys [type field]} (get-in query-rec [:projections column])]
-              (if value
-                (case type
-                  :multi
-                  (map->BinaryExpression {:operator (keyword op)
-                                          :column (columns->fields ["value_integer" "value_float"])
-                                          :value (if (number? value) [value value]
-                                                     (map ks/parse-number [value value]))})
-
-                  (map->BinaryExpression {:operator (keyword op)
-                                          :column field
-                                          :value  (if (= :timestamp type)
-                                                    (to-timestamp value)
-                                                    (ks/parse-number (str value)))}))
-                (throw (IllegalArgumentException.
-                        (format "Value %s must be a number for %s comparison." value op)))))
+              (if (or (= :timestamp type) (and (= :number type) (not (string? value))))
```

Comment:
type refers to the column type. value refers to the user-supplied argument. This handles queries like ["<", "line", "20"] which we should be rejecting.